fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryManager.java

```diff
@@ -122,51 +127,52 @@ public CompletableFuture<Void> mergeIndex(@Nonnull LucenePartitioner partitioner
                 null,
                 ScanProperties.FORWARD_SCAN);
 
-        // TO DO: https://github.com/FoundationDB/fdb-record-layer/issues/2414
         return cursor
                 .map(tuple ->
                         Tuple.fromItems(tuple.getItems().subList(0, groupingCount)))
                 .forEachAsync(groupingKey -> mergeIndex(analyzerWrapper, groupingKey, partitioner, agilityContext),
-                        2);
+                        2)
+                .whenComplete((vignore, eignore) -> agilityContext.flushAndClose());
     }
 
     private CompletableFuture<Void> mergeIndex(LuceneAnalyzerWrapper analyzerWrapper, Tuple groupingKey,
                                                @Nonnull LucenePartitioner partitioner, final AgilityContext agileContext) {
         if (!partitioner.isPartitioningEnabled()) {
-            try {
-                getDirectoryWrapper(groupingKey, null, agileContext).mergeIndex(analyzerWrapper);
-                return AsyncUtil.DONE;
-            } catch (IOException e) {
-                throw new RecordCoreStorageException("Lucene mergeIndex failed", e)
-                        .addLogInfo(LuceneLogMessageKeys.GROUP, groupingKey);
-            }
+            mergeIndexNow(analyzerWrapper, groupingKey, null);
```

Comment:
Do we want this to be using the same `agileContext` as the one used for getting the grouping keys?
In the old code, it reused the `AgileContext`, but with the new code, it creates a separate `AgileContext` for each merge.
It seems as though this could be creating an excessive number of transactions in the case where most groups/partitions don't need to do a merge. If you can open 10 indexes in one transaction, then the old code could go through 10 partitions in a single context, but this will now create at least 11 (one for grouping keys & partition metadata and one for each partition)

The upside to the approach here is that if you have a merge that actually does work on a single partition, you can be sure that work is committed, and decrease the chances of being half-way through a merge. Also, it decreases the chance that you will do one merge, and fail on the second merge, and lose all the work from the first, because writing the `segments_N` ended up in the second transaction. And, if you have partitions, each partition is probably getting close enough to the limit that opening more than a couple probably won't work anyways.


I'm not sure that I have a clear answer as to which is a better approach for efficient resource utilization, so going with this change is probably reasonable, as it is a simple solution to the leaking of contexts.
I just want to make sure that we are making a conscious decision here.